apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemDAO.java

```diff
@@ -379,47 +388,49 @@ public Mono<Long> save(@NonNull UUID datasetId, @NonNull List<DatasetItem> items
             return Mono.empty();
         }
 
-        return inset(datasetId, items)
-                .retryWhen(AsyncUtils.handleConnectionError());
+        return inset(datasetId, items);
     }
 
     private Mono<Long> inset(UUID datasetId, List<DatasetItem> items) {
-        return asyncTemplate.nonTransaction(connection -> {
+        List<List<DatasetItem>> batches = Lists.partition(items, bulkConfig.getSize());
 
-            var statement = connection.createStatement(INSERT_DATASET_ITEM);
-
-            return mapAndInsert(datasetId, items, statement)
-                    .flatMap(Result::getRowsUpdated)
-                    .reduce(0L, Long::sum);
-        });
+        return Flux.fromIterable(batches)
+                .flatMapSequential(batch -> asyncTemplate.nonTransaction(connection -> mapAndInsert(datasetId, batch, connection)))
+                .reduce(0L, Long::sum);
     }
 
-    private Flux<? extends Result> mapAndInsert(UUID datasetId, List<DatasetItem> items, Statement statement) {
-        return makeFluxContextAware((userName, workspaceName, workspaceId) -> {
-
-            for (Iterator<DatasetItem> iterator = items.iterator(); iterator.hasNext();) {
-                var item = iterator.next();
-
-                statement.bind("id", item.id())
-                        .bind("datasetId", datasetId)
-                        .bind("input", item.input().toString())
-                        .bind("source", item.source().getValue())
-                        .bind("traceId", getOrDefault(item.traceId()))
-                        .bind("spanId", getOrDefault(item.spanId()))
-                        .bind("expectedOutput", getOrDefault(item.expectedOutput()))
-                        .bind("metadata", getOrDefault(item.metadata()))
-                        .bind("workspace_id", workspaceId)
-                        .bind("createdBy", userName)
-                        .bind("lastUpdatedBy", userName);
-
-                if (iterator.hasNext()) {
-                    statement.add();
-                }
-            }
+    private Mono<Long> mapAndInsert(UUID datasetId, List<DatasetItem> items, Connection connection) {
+
+        List<QueryItem> queryItems = getQueryItemPlaceHolder(items);
 
-            statement.fetchSize(items.size());
+        var template = new ST(INSERT_DATASET_ITEM)
+                .add("items", queryItems);
 
-            return Flux.from(statement.execute());
+        String sql = template.render();
+
+        var statement = connection.createStatement(sql);
```

Comment:
We should try a `createBatch` call and work from there.

Reply:
As discussed over Slack, The problem with the batch implementation from R2dbc for batch is that it only accepts plain SQL strings. There is no binding for parameters; using string templates will open the door for SQL injections. That's why I implemented like this

Reply:
Discussed offline. The support in the R2DBC is very limited as only allows to pass SQL statements as Strings, without the possibility of binding.